core/src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java

```diff
@@ -556,11 +512,135 @@ void toSerializable(DistributedStorage storage)
     parallelExecutor.validate(tasks, getId());
   }
 
-  private TableMetadata getTableMetadata(Scan scan) throws ExecutionException {
-    TransactionTableMetadata metadata = tableMetadataManager.getTransactionTableMetadata(scan);
+  /**
+   * Validates the scan results to check if there is no anti-dependency.
+   *
+   * <p>This method scans the latest data and compares it with the scan results. If there is a
+   * discrepancy, it means that the scan results are changed by another transaction. In this case,
+   * an {@link ValidationConflictException} is thrown.
+   *
+   * <p>Since the validation is performed after the prepare-record phase, the scan might include
+   * prepared records if the transaction has performed operations that affect the scan result. In
+   * such cases, those prepared records can be safely ignored.
+   *
+   * <p>Note that this logic is based on the assumption that identical scans return results in the
+   * same order, provided that the underlying data remains unchanged.
+   *
+   * @param storage a distributed storage
+   * @param scan the scan to be validated
+   * @param results the results of the scan
+   * @throws ExecutionException if a storage operation fails
+   * @throws ValidationConflictException if the scan results are changed by another transaction
+   */
+  private void validateScanResults(
+      DistributedStorage storage, Scan scan, LinkedHashMap<Key, TransactionResult> results)
+      throws ExecutionException, ValidationConflictException {
+    Scanner scanner = null;
+    try {
+      // Only get tx_id, tx_version and primary key columns because we use only them to compare
+      scan.clearProjections();
+      scan.withProjection(Attribute.ID).withProjection(Attribute.VERSION);
+      ScalarDbUtils.addProjectionsForKeys(scan, getTableMetadata(scan));
+
+      if (scan.getLimit() == 0) {
+        scanner = storage.scan(scan);
+      } else {
+        // Get a scanner without the limit if the scan has a limit
+        scanner = storage.scan(Scan.newBuilder(scan).limit(0).build());
+      }
+
+      //
+      // Initialize the iterators for the scan results and the latest scan results
+      //
+      Optional<Result> latestResult = scanner.one();
+
+      Iterator<Entry<Key, TransactionResult>> resultsIterator = results.entrySet().iterator();
+      Entry<Key, TransactionResult> resultsEntry = Iterators.getNext(resultsIterator, null);
+
+      // Compare the scan results with the latest scan results
+      while (latestResult.isPresent() && resultsEntry != null) {
+        TransactionResult latestTxResult = new TransactionResult(latestResult.get());
+        Key key = new Key(scan, latestTxResult);
+
+        if (latestTxResult.getId() != null && latestTxResult.getId().equals(id)) {
+          // The record is inserted/deleted/updated by this transaction
```

Comment:
I might miss something, but the result has been retrieved, so it cannot be deleted by this transaction?

Reply:
Thank you for the explanation offline! I understand the logic.